### ISSUE TITLE
TestLexers can't associate more than one file extension per module

### ISSUE THREAD

**Issue by rdipardo** [ScintillaOrg/lexilla#81]
Lexilla's [test runner] can't parse a `lexer.*` key containing multiple file extensions, in the manner of a `file.patterns` definition: `lexer.*.ext1;*.ext2;*.ext3=lexmodule`

While acknowledging the runner's necessarily limited functionality, I think even a rudimentary substitute for `file.patterns` may prove useful when testing lexers that specify multiple languages, or mutually incompatible dialects and standards of a single language. A suite of test files with various extensions could then share a common set of static properties, optionally supplemented by [conditional ones].

For example, a downstream application may want to implement an option suppressing JSON-LD constructs in plain JSON files. A contributor writes a pair of test files with similar content, one with the [IANA-designated] `*.jsonld` extension, the other ending in `*.json`. A single `*.properties` file would be sufficient if both extensions could be mapped to the JSON lexer:

~~~
lexer.*.json;*.jsonld=json

# common properties
fold=1
# etc.

match x.jsonld
    # properties unique to JSON-LD

match x.json
    # properties unique to JSON
~~~

This patch extends `PropertyMap::GetPropertyForFile` to also check for a delimited list of file extensions:  [0001-TestLexers-Multi-File-Exts.diff.txt](https://github.com/ScintillaOrg/lexilla/files/8595636/0001-TestLexers-Multi-File-Exts.diff.txt)

[test runner]: https://github.com/ScintillaOrg/lexilla/tree/master/test#readme
[conditional ones]: https://github.com/ScintillaOrg/lexilla/commit/1499a42c12e454e8d6e3c1a6ada1ddc3a52e0407
[IANA-designated]: https://json-ld.org/spec/FCGS/json-ld/20180607/#:~:text=File%20extension(s),-%3A

**Comment by nyamatongwe** [comment]
TestLexers can currently use multiple lexer statements:
~~~text
lexer.*.py=python
lexer.*.pxx=python
~~~

Adding file patterns like the patch should help those familiar with SciTE.

**Comment by rdipardo** [comment]
> TestLexers can currently use multiple lexer statements

I would have assumed each key gets read only once, but that works, too.
In fact, the test suite under the `hypertext` directory successfully uses a global wildcard:

https://github.com/ScintillaOrg/lexilla/blob/b7b401a2b169aa58dffa96888aca96b38c292a3c/test/examples/hypertext/SciTE.properties#L1

I suppose the issue's title is a bit misleading. What's really needed is a concise way to specify 2 or more extensions. Currently it's a choice between one or arbitrarily many, or else duplicating shared property keys for each extension.

**Comment by nyamatongwe** [comment]
The patch may throw if there is no '.' in a part as `part.find_first_of('.')` returns `npos` which isn't good to pass on to `substr`. SciTE allows non `*.` in parts, for example, `makefile;Makefile;*.mak;configure` which can fail to work but shouldn't crash.

The `*` is currently omitted from `keySuffix` so may be `.py;*.pxx` which is irregular - it may be better to include the initial `*` to test against `*.py;*.pxx` with a common testing function between the single and multi-part code.

Success can return early instead of breaking the loop which collapses the scope of `match`.

**Comment by rdipardo** [comment]
Sorry to reply with a more additive patch, but it wasn't long before the required validation logic was enough to fill a method.
The previously added branch has been reduced to a call site, and the parsing routines now take the entire property key into account, not just a substring. This new attempt does real validation, prints a warning message, and hasn't thrown any exceptions for me yet: 

[0002-Issue81-Validate-File-Patterns.diff.txt](https://github.com/ScintillaOrg/lexilla/files/8601217/0002-Issue81-Validate-File-Patterns.diff.txt)

Simply avoiding the out-of-range exception was always achievable by passing `0` as the first argument to the affected `substr` call. Less tractable was the naïvely permissive parsing logic.

Even when non-fatal, a missing period in the first file pattern of a list would still result in a match. Given:

    lexer.*html;*.php;*.asp=hypertext

the resulting list was:

    html
    .php
    .asp

A missing period at any later position made the nearest extension invisible. Given:

    lexer.*.html;*php;*.asp=hypertext

the list was:

    .html
    ''
    .asp

Similarly,

    lexer.*.html;*.php;*asp=hypertext

    .html
    .php
    ''

An [undocumented][0] behaviour of `std::string::ends_with` seems to be that it returns true when given an empty string (or maybe just null-terminated C strings?), so malformed file patterns were bound to cause false positives.



As a demonstration and regression check, here is a contrived test case using file patterns to make all comments errors in JSON-LD files but not JSON: 

[0003-Issue81-POC-lexer-test.txt](https://github.com/ScintillaOrg/lexilla/files/8601223/0003-Issue81-POC-lexer-test.txt)

[0]: https://en.cppreference.com/w/cpp/string/basic_string/ends_with

**Comment by nyamatongwe** [comment]
There shouldn't be any `=` in keys so `key.find_last_of('=')` shouldn't be needed.

~~~c++
// Only accept patterns in the form *.xyz
~~~

If that is accurate then the checks can be simplified to `pat.starts_with("*.")`.

**Comment by rdipardo** [comment]
> There shouldn't be any `=` in keys so `key.find_last_of('=')` shouldn't be needed.

Corrected.

> checks can be simplified to `pat.starts_with("*.")`.

It increases the required standard to C++20, but I see now that's been the case [for some time already][0]. I apparently thought that "find_first_of" meant "first **instance** of".

Requiring the whole prefix is better but prevents accurately diagnosing a missing '.' as the branch is skipped if any part is absent. As a compromise, this new patch checks with `starts_with` and prints a more general diagnostic at a new `else `branch:

[0004-Issue81-Optimize-File-Patterns-Validation.diff.txt](https://github.com/ScintillaOrg/lexilla/files/8608135/0004-Issue81-Optimize-File-Patterns-Validation.diff.txt)

This file contains only the changes since the previous patch: 

[0004-Issue81-deltas.txt](https://github.com/ScintillaOrg/lexilla/files/8608134/0004-Issue81-deltas.txt)


It predates this issue, but MSVC is raising an obscure diagnostic apropos this concatenation:

https://github.com/ScintillaOrg/lexilla/blob/ff879c5dd2e46e3c08968e025d37c36ec80f13b0/test/TestLexers.cxx#L705

    warning C4244: 'argument': conversion from 'int' to 'const _Elem', possible loss of data
            with
            [
                _Elem=char
            ]


[0]: https://github.com/ScintillaOrg/lexilla/tree/master/test#readme

**Comment by rdipardo** [comment]
More tidying up for a net difference of &#x2013;5 lines:

- no longer search the indices of '*' and '.' now that `starts_with` guarantees their position
- only assert that the pattern is longer than the prefix as that's now sufficient
- slice a `const` local string instead of deleting bytes from a mutable one
- declare `lexerPrefix` as `constexpr std::string_view` to match the other prefixes (does not appease warning C4244, however)

[0005-Issue81-Further-Optimize-File-Patterns-Validation.diff.txt](https://github.com/ScintillaOrg/lexilla/files/8608491/0005-Issue81-Further-Optimize-File-Patterns-Validation.diff.txt)

Just the differences: [0005-Issue81-deltas.txt](https://github.com/ScintillaOrg/lexilla/files/8608454/0005-Issue81-deltas.txt)

**Comment by nyamatongwe** [comment]
I don't see C4244 with MSVC 2022 in c++latest or c++20 modes for 64-bit or 32-bit builds.

The change to `TestLexers.cxx` looks good.

Is `Issue81.jsonld` worth including as a test case for the json lexer?

**Comment by rdipardo** [comment]
> I don't see C4244 with MSVC 2022 in c++latest or c++20 modes for 64-bit or 32-bit builds.

It must be a level 4 warning; MSVC 2019 raises it when invoked with -Zi -TP -MP -W4 -Od -MTd -DDEBUG -EHsc -std:c++latest

> Is `Issue81.jsonld` worth including as a test case for the json lexer?

It's pretty trivial and could stymie the innovation of a truly useful optional lexer property in the future.
The file patterns feature should be discoverable, not something contributors will bump into it.

As for the implementation, it seems now that searching for a semi-colon delimiter at the top of the method is redundant when the callee does the same before the method runs. This consolidated patch reduces overall insertions and conserves horizontal space:

[0006-Less-Redundant-File-Patterns-Check.diff.txt](https://github.com/ScintillaOrg/lexilla/files/8635521/0006-Less-Redundant-File-Patterns-Check.diff.txt)

**Comment by rdipardo** [comment]
> I don't see C4244 with MSVC 2022 in c++latest or c++20 modes for 64-bit or 32-bit builds.

While I can't sincerely recommend it, this kludge appears to silence MSVC:

~~~diff
diff --git a/test/TestLexers.cxx b/test/TestLexers.cxx
index 7c777dfb..1360c4fa 100644
--- a/test/TestLexers.cxx
+++ b/test/TestLexers.cxx
@@ -699,10 +699,11 @@ void SetProperties(Scintilla::ILexer5 *plex, const PropertyMap &propertyMap, std
 	assert(plex);
 
 	// Set keywords, keywords2, ... keywords9, for this file
-	for (int kw = 0; kw < 10; kw++) {
+	const char *kwIDs[8] = { "2", "3", "4", "5", "6", "7", "8", "9" };
+	for (int kw = 0; kw < 9; kw++) {
 		std::string kwChoice("keywords");
 		if (kw > 0) {
-			kwChoice.push_back('1' + kw);
+			kwChoice.append(kwIDs[kw - 1]);
 		}
 		kwChoice.append(".*");
 		std::optional<std::string> keywordN = propertyMap.GetPropertyForFile(kwChoice, fileName);

~~~

**Comment by nyamatongwe** [comment]
Included in 5.1.7 release.
